lib/grpnames.gi

```diff
@@ -709,56 +710,137 @@ InstallGlobalFunction(DirectFactorsOfGroupKN,
+    G := arg[1];
+    method := arg[2];
+    fullNs := true;
+  elif Length(arg) = 2 and IsGroup(arg[1]) and IsList(arg[2])
+      and ForAll( Set(arg[2]), N ->
+                              IsSubgroup(arg[1], N) and IsNormal(arg[1], N))
+      then
+    G := arg[1];
+    Ns := ShallowCopy(arg[2]);
+    fullNs := false;
+  elif Length(arg) = 3 and IsGroup(arg[1]) and IsList(arg[2])
+      and ForAll( Set(arg[2]), N ->
+                              IsSubgroup(arg[1], N) and IsNormal(arg[1], N))
+      and arg[3] in ["all", "any", "str"] then
+    G := arg[1];
+    Ns := ShallowCopy(arg[2]);
+    method := arg[3];
+    fullNs := false;
+  else
+    Error("usage: SemidirectDecompositionsOfFiniteGroup(<G> [, <Ns>] [, <mthd>])");
+  fi;
 
-    # non-1-or-G normal subgroups
-    Ns := Reversed( Filtered(Hs, H -> IsNormal(G, H)) );
+  if HasSemidirectDecompositions(G) then
+    NHs := [ ];
+    for NH in SemidirectDecompositions(G) do
+      N := NH[1];
+      H := NH[2];
+      if method in [ "any", "str" ] and not IsTrivial(N)
+        and not IsTrivial(H) and
+        ( not IsBound(Ns) or N in Ns ) then
+        return [ N, H ];
+      elif method="all" and
+        ( not IsBound(Ns) or N in Ns ) then
+        AddSet(NHs, [ N, H ]);
+      fi;
+    od;
+    if method in [ "any", "str" ] then
+      return fail;
+    elif method = "all" then
+      return NHs;
+    fi;
+  fi;
 
-    # find first decomposition
-    firstHN := function ()
+  if method in [ "any", "str" ] then
+    if HasSemidirectProductInfo(G) then
+      N := Image(Embedding(G, 2));
+      H := Image(Embedding(G, 1));
+      if not IsTrivial(N) and not IsTrivial(H) and
+        ( not IsBound(Ns) or N in Ns ) then
+        return [ N, H ];
+      fi;
+    fi;
+    N := NormalHallSubgroupsFromSylows(G, "any");
+    if N <> fail then
+      # by the Schur-Zassenhaus theorem there must exist a complement
+      if method = "any" then
+        H := ComplementClassesRepresentatives(G, N)[1];
+        return [ N, H ];
+      # only the isomorphism type of the complement is interesting
+      elif method = "str" then
+        Assert(1, Length( ComplementClassesRepresentatives(G, N) ) > 0);
+        return [ N, G/N ];
+      fi;
+    fi;
+  fi;
 
-      local H, N, sizeNs;
+  # simple groups have no nontrivial normal subgroups
+  if IsSimpleGroup(G) then
+    if method in [ "any", "str" ] then
+      return fail;
+    elif method = "all" then
+      return [ [TrivialSubgroup(G), G], [G, TrivialSubgroup(G)] ];
+    fi;
+  fi;
 
-      sizeNs := List(Ns, Size);
-      for H in Hs do
-        if Size(G)/Size(H) in sizeNs then
-          for N in Filtered(Ns, N -> Size(N) = Size(G)/Size(H)) do
-            if IsTrivial(NormalIntersection(N, H)) then
-              return Size(H);
-            fi;
-          od;
-        fi;
-      od;
-      return 0;
-    end;
+  if not IsBound(Ns) then
+    Ns := ShallowCopy(NormalSubgroups(G));
```

Comment:
@fingolfin @alex-konovalov I have tried it with -r -A -x 80 -m 100m -o 1g. It is not enough. For one, ````SmallGroup(16, 13)```` (and a lot of other groups of size <100) have different ````StructureDescription```` with and without packages. I could try to sort the ````NormalSubgroups```` completely according to some general way, but that would defeat efficiency. And I am not entirely sure how to do it, either. 

Second, in StructureDescription.tst:25 the group will have ````A5 x A5```` both with and without packages, but if no packages are present then maxsub.gi:545 will be invoked, and an info message will be printed along with the output. Not sure if that can be nicely reconciled.

Finally, _without_ packages the test runs in 8 seconds on my laptop, _with_ packages it runs for 15 seconds. I am wondering if this is really how it should be....

I am not sure how to proceed at this point. 

PS. Now that #563 is merged, I think I should rebase according to new master and uncomment the smallest example ````A5 : S5```` in the tests. Would that be ok, or should I wait until the current problems are sorted?